### PR TITLE
feat(rover): implement supergraph

### PR DESCRIPTION
**Pull request by leeroyrose**

### CODE REVIEW

File: json.d.ts [hunk 1]
```diff
@@ -1,4 +0,0 @@
-declare module '*.json' {
```
Comment (matt-halliday) [inline json.d.ts:1]: I thought this seemed dumb but I think I fixed the issue I was having and just never got rid of it 👍

File: packages/gateway/package.json [hunk 2]
```diff
@@ -3,12 +3,10 @@
...
     "main": "index.js",
     "scripts": {
+        "rover": "rover supergraph compose --config ./supergraph-config.yaml > supergraph.graphql",
         "test": "echo \"Error: no test specified\" && exit 1"
     },
     "author": "",
     "license": "ISC",
-    "description": "",
-    "dependencies": {
-        "@apollo/gateway": "^0.41.0"
```
Comment (matt-halliday) [inline packages/gateway/package.json]: I was wondering about whether these gateway/rover deps should be specific to this package - having not done much with workspaces yet, is there a best practice for where to include deps in a workspace set up?

File: packages/weather/index.ts [hunk 3]
```diff
@@ -1,18 +1,14 @@
...
 import weatherData from './weather.json';
-import { buildServer } from 'lib';
-import * as console from 'console';
-
-const typeDefs = gql`
-    type Weather @key(fields: "code") {
-        code: ID
-        name: String
-    }
+import { buildServer } from '../lib';
```
Comment (matt-halliday) [inline packages/weather/index.ts:2]: My IDE is telling me to shorten this to just `'lib'`. I'm guessing it's because of the workspace set up, it can view it as a package.

File: packages/city/city.graphql [hunk 4]
```diff
@@ -0,0 +1,14 @@
+type City {
+    name: String
+    weatherCode: ID
+    currentWeather: Weather
+}
+
+extend type Weather @key(fields: "code") {
```
Comment (matt-halliday) [inline packages/city/city.graphql:7]: My IDE is complaining that this directive is not found, similarly complaining in lots of places that stuff is being redefined as a result of the supergraph schema. I wonder if there's a way to make everyone happy in a workspace set up.

File: packages/gateway/index.ts [hunk 5]
```diff
@@ -1,54 +1,20 @@
-import { ApolloServer, gql } from 'apollo-server';
+import { ApolloServer } from 'apollo-server';
 import { ApolloGateway } from '@apollo/gateway';
-import * as console from 'console';
+import * as fs from 'fs';
+import * as path from 'path';
 
-gql`
```
Comment (matt-halliday) [inline packages/gateway/index.ts:5]: This blob here was funny. I wasn't using it anywhere but it made my IDE happy. I think this whole static analysis/IDE pleasing side of things will take some hacking, but I hope it doesn't

File: packages/gateway/package.json [hunk 6]
```diff
@@ -3,12 +3,10 @@
     "version": "1.0.0",
     "main": "index.js",
     "scripts": {
+        "rover": "rover supergraph compose --config ./supergraph-config.yaml > supergraph.graphql",
```
Comment (matt-halliday) [inline packages/gateway/package.json:6]: This is the bit that I spent 2+ hours banging my head against... and what you have done is the same. So I am super confused, but glad it's working 🥴

**Review by matt-halliday — APPROVED** [review]
wunderbar!

File: packages/gateway/index.ts [hunk 7]
```diff
@@ -1,54 +1,20 @@
...
 
-(async () => {
-    const { schema, executor } = await gateway.load();
-
-    const server = new ApolloServer({
-        schema,
-        executor,
-    });
+const server = new ApolloServer({
+    gateway,
```
Comment (matt-halliday) [inline packages/gateway/index.ts:16]: I ended up adding the following to this so I could play around in the browser locally.
```suggestion
    gateway,
    plugins: [ApolloServerPluginLandingPageGraphQLPlayground()],
```

File: json.d.ts (hunk 1 above)
Comment (leeroyrose) [reply json.d.ts:1]: I thought so, all good!

File: packages/gateway/package.json (hunk 2 above)
Comment (leeroyrose) [reply packages/gateway/package.json]: Yeah I think all shared at top level and all specific in nested packages, so i think i've done the wrong thing here, it should have stayed. I'll re-work it

File: packages/weather/index.ts (hunk 3 above)
Comment (leeroyrose) [reply packages/weather/index.ts:2]: Strange, mine didn't do that at all. I just did this to get it compiling. Did you do anything specific in your ide to setup the workspaces?

File: packages/city/city.graphql (hunk 4 above)
Comment (leeroyrose) [reply packages/city/city.graphql:7]: Might be an IDE specific thing, VSCode with the graphql plugins and it's all good. What ide are you using?

File: packages/gateway/index.ts (hunk 7 above)
Comment (leeroyrose) [reply packages/gateway/index.ts:16]: Ah cool, you can use the graphql playground locally which is probably a bit nicer than the browser

https://github.com/graphql/graphql-playground

File: packages/city/city.graphql (hunk 4 above)
Comment (matt-halliday) [reply packages/city/city.graphql:7]: intellij IDEA

File: packages/gateway/index.ts [hunk 8]
```diff
@@ -1,54 +1,21 @@
-import { ApolloServer, gql } from 'apollo-server';
+import { ApolloServer } from 'apollo-server';
 import { ApolloGateway } from '@apollo/gateway';
-import * as console from 'console';
+import * as fs from 'fs';
+import * as path from 'path';
```
Comment (matt-halliday) [inline packages/gateway/index.ts:4]: For the below suggestion:
```suggestion
import * as path from 'path';
import { ApolloServerPluginLandingPageGraphQLPlayground } from 'apollo-server-core';
```

File: packages/weather/index.ts (hunk 3 above)
Comment (matt-halliday) [reply packages/weather/index.ts:2]: no - I didn't touch it, it seems to see the `package.json` files from the other packages and use the package names as if they were node dependencies.

Comment (matt-halliday) [reply packages/weather/index.ts:2]: it works this way too - it just gives me a syntax suggestion to shorten it

Comment (matt-halliday) [reply packages/weather/index.ts:2]: Screenshot 2021-09-15 at 11 51 17

File: packages/gateway/index.ts (hunk 5 above)
Comment (matt-halliday) [reply packages/gateway/index.ts:5]: perhaps a `shim.graphql` might please the gods

File: packages/weather/index.ts (hunk 3 above)
Comment (leeroyrose) [reply packages/weather/index.ts:2]: I think usually you would pull this in as it's own npm package rather than reaching into it cross directory. For the sake of this, we can reach but yeah I don't know how that syntax worked for you! My ide hated it lol
